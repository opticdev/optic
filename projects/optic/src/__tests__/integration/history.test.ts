import { jest, describe, test, expect } from '@jest/globals';
import {
  runOptic,
  setupWorkspace,
  normalizeWorkspace,
  run,
} from './integration';
jest.setTimeout(30000);

const date = new Date('2023-01-01');

describe('optic history', () => {
  test('writes changelog', async () => {
    process.env.OPTIC_TOKEN = 'something';
    const workspace = await setupWorkspace('history/petstore', {
      repo: true,
      commit: true,
    });
    const commitDateChanged = formatDateForGitCommit(date);

    const command = `
cp petstore-updated.json petstore-base.json &&\
git add . &&\
GIT_COMMITTER_DATE="${commitDateChanged}" git commit -m 'update petstore' &&\
mv petstore-base.json petstore-base-renamed.json &&\
git add . &&\
GIT_COMMITTER_DATE="${commitDateChanged}" git commit -m 'rename petstore' &&\
touch abc.def && git add . &&\
GIT_COMMITTER_DATE="${formatDateForGitCommit(
      new Date('2023-01-02')
    )}" git commit -m 'empty-commit'
`;

    await run(command, false, workspace);

    const { combined, code } = await runOptic(
      workspace,
      'history petstore-base-renamed.json'
    );

    expect(code).toBe(0);
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
  });

  test('no changes', async () => {
    const workspace = await setupWorkspace('history/petstore', {
      repo: true,
      commit: true,
    });

    const { combined, code } = await runOptic(
      workspace,
      'history petstore-updated.json'
    );

    expect(code).toBe(0);
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
  });
});

function formatDateForGitCommit(date: Date) {
  const year = date.getUTCFullYear();
  const month = padZero(date.getUTCMonth() + 1);
  const day = padZero(date.getUTCDate());
  const hours = padZero(date.getUTCHours());
  const minutes = padZero(date.getUTCMinutes());
  const seconds = padZero(date.getUTCSeconds());
  const offset = getTimezoneOffset(date.getTimezoneOffset());

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds} ${offset}`;
}

function padZero(value: number) {
  return value.toString().padStart(2, '0');
}

function getTimezoneOffset(offset: number) {
  const sign = offset > 0 ? '-' : '+';
  const absOffset = Math.abs(offset);
  const hours = padZero(Math.floor(absOffset / 60));
  const minutes = padZero(absOffset % 60);

  return `${sign}${hours}${minutes}`;
}
