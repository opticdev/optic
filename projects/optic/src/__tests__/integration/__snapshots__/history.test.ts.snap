// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optic history no changes 1`] = `
"# Swagger Petstore Updated
No changes found
"
`;

exports[`optic history writes changelog 1`] = `
"# Swagger Petstore Updated
### Sun Jan 01 2023
- \`PUT\` \`/user/{username}\`:
  - \`requestBody\`: removed support for \`phone\` property
  - \`requestBody\`: added support for \`bio\` property
  - \`requestBody\`: changed \`userStatus\`.\`type\` to \`string\`
  - \`requestBody\`: removed \`userStatus\`.\`format\`
  - \`requestBody\`: added \`userStatus\`.\`enum\`
  - \`requestBody\`: \`lastName\` is no longer required
  - \`requestBody\`: \`username\` is now required
- \`GET\` \`/user/{username}\`:
  - \`404\` response: added \`application/json\`
  - \`200\` response: removed support for \`phone\` property
  - \`200\` response: added support for \`bio\` property
  - \`200\` response: changed \`userStatus\`.\`type\` to \`string\`
  - \`200\` response: removed \`userStatus\`.\`format\`
  - \`200\` response: added \`userStatus\`.\`enum\`
  - \`200\` response: \`lastName\` is no longer required
  - \`200\` response: \`username\` is now required
- \`GET\` \`/user/login\`:
  - \`200\` response: removed \`headers\`.\`content-type\`
  - \`200\` response: added \`headers\`.\`X-Rate-Limit\`
  - \`200\` response: changed \`headers\`.\`X-Expires-After\`.\`type\` to \`string\`
  - \`200\` response: added \`headers\`.\`X-Expires-After\`.\`format\`
- \`POST\` \`/user/createWithList\`:
  - \`requestBody\`: removed support for \`items\`.\`phone\` property
  - \`requestBody\`: added support for \`items\`.\`bio\` property
  - \`requestBody\`: changed \`items\`.\`userStatus\`.\`type\` to \`string\`
  - \`requestBody\`: removed \`items\`.\`userStatus\`.\`format\`
  - \`requestBody\`: added \`items\`.\`userStatus\`.\`enum\`
  - \`requestBody\`: \`items\`.\`lastName\` is no longer required
  - \`requestBody\`: \`items\`.\`username\` is now required
- \`POST\` \`/user/createWithArray\`:
  - \`requestBody\`: removed support for \`items\`.\`phone\` property
  - \`requestBody\`: added support for \`items\`.\`bio\` property
  - \`requestBody\`: changed \`items\`.\`userStatus\`.\`type\` to \`string\`
  - \`requestBody\`: removed \`items\`.\`userStatus\`.\`format\`
  - \`requestBody\`: added \`items\`.\`userStatus\`.\`enum\`
  - \`requestBody\`: \`items\`.\`lastName\` is no longer required
  - \`requestBody\`: \`items\`.\`username\` is now required
- \`POST\` \`/user\`:
  - \`requestBody\`: removed support for \`phone\` property
  - \`requestBody\`: added support for \`bio\` property
  - \`requestBody\`: changed \`userStatus\`.\`type\` to \`string\`
  - \`requestBody\`: removed \`userStatus\`.\`format\`
  - \`requestBody\`: added \`userStatus\`.\`enum\`
  - \`requestBody\`: \`lastName\` is no longer required
  - \`requestBody\`: \`username\` is now required
- \`DELETE\` \`/store/order/{orderId}\`: \`404\` response: added \`application/json\`
- \`GET\` \`/store/order/{orderId}\`:
  - \`404\` response: added \`application/json\`
  - \`200\` response: added support for \`summary\` property
  - \`200\` response: removed support for value \`approved\` from enum \`status\`
  - \`200\` response: added support for new value \`canceled\` on enum \`status\`
  - \`200\` response: \`petId\` is no longer required
- \`POST\` \`/store/order\`:
  - \`200\` response: added support for \`summary\` property
  - \`200\` response: removed support for value \`approved\` from enum \`status\`
  - \`200\` response: added support for new value \`canceled\` on enum \`status\`
  - \`requestBody\`: added support for \`summary\` property
  - \`requestBody\`: removed support for value \`approved\` from enum \`status\`
  - \`requestBody\`: added support for new value \`canceled\` on enum \`status\`
- \`POST\` \`/pet/{petId}/uploadImage\`:
  - \`200\` response: removed 
  - \`201\` response: added 
  - \`404\` response: added 
  - \`default\` response: changed \`type\`.\`type\` to \`integer\`
  - added \`debug\` cookie parameter
- removed \`GET\` \`/pet\`
- \`PUT\` \`/pet\`:
  - \`400\` response: removed 
  - \`requestBody\`: removed 
  - \`requestBody\`: removed support for \`name\` property
  - \`requestBody\`: added support for \`number\` property
  - \`requestBody\`: \`name\` is no longer required
  - \`requestBody\`: \`number\` is now required
- \`GET\` \`/parameters_at_path\`: added \`query_in_path\` query parameter
- \`POST\` \`/example\`: removed \`somethingelse\` query parameter
- \`GET\` \`/example\`:
  - \`200\` response: changed \`toTypeArrays\`.\`type\`
  - \`200\` response: changed \`fromTypeArrays\`.\`type\` to \`string\`
  - \`200\` response: \`requiredKeys\`.\`requiredToOptional\` is no longer required
  - \`200\` response: \`requiredKeys\`.\`deletedRequiredKey\` is no longer required
  - \`200\` response: \`requiredKeys\`.\`optionalToRequired\` is now required
  - \`200\` response: \`requiredKeys\`.\`addedRequiredKey\` is now required
  - \`200\` response: removed support for \`requiredKeys\`.\`deletedRequiredKey\` property
  - \`200\` response: added support for \`requiredKeys\`.\`addedRequiredKey\` property
  - \`200\` response: added an item of type \`string\` to \`expandableObject\` \`anyOf\` values
  - \`200\` response: changed \`expandableObject\`.\`orderId\`.\`type\` to \`string\`
  - \`200\` response: changed \`composedObject\`.\`orderId\`.\`type\` to \`number\`
  - \`200\` response: removed an item of type \`object\` from \`stringOrNumberOrObject\` \`oneOf\` values
  - removed \`deletedQueryParameter\` query parameter
  - added \`addedQueryParameter\` query parameter
  - removed \`requiredToOptional\` query parameter
  - added \`optionalToRequired\` query parameter
- added \`GET\` \`/pet/findByStatus\`
- added \`GET\` \`/pet/findByTags\`
- added \`GET\` \`/pet/{petId}\`
- added \`POST\` \`/pet/{petId}\`
- added \`DELETE\` \`/pet/{petId}\`
"
`;
